pkg/apiserver/debugapi/client.go

```diff
@@ -66,36 +98,69 @@ func buildRelativeURI(path string, query string) string {
 
 type tidbImplement struct {
 	fx.In
-	Client *tidb.Client
+	Client     *tidb.Client
+	EtcdClient *clientv3.Client
 }
 
 func (impl *tidbImplement) Get(req *endpoint.Request) (*httpc.Response, error) {
 	return impl.Client.
 		WithEnforcedStatusAPIAddress(req.Host, req.Port).
-		WithStatusAPITimeout(defaultTimeout).
-		Get(buildRelativeURI(req.Path, req.Query))
+		WithStatusAPITimeout(req.Timeout).
+		Get(buildRelativeURI(req.Path(), req.Query()))
+}
+
+func (impl *tidbImplement) CheckDest(host string, port int) (bool, error) {
+	info, err := topology.FetchTiDBTopology(context.Background(), impl.EtcdClient)
+	if err != nil {
+		return false, err
+	}
+	return funk.Contains(info, func(item topology.TiDBInfo) bool {
+		return item.IP == host && item.StatusPort == uint(port)
+	}), nil
 }
 
 type tikvImplement struct {
 	fx.In
-	Client *tikv.Client
+	Client   *tikv.Client
+	PDClient *pd.Client
 }
 
 func (impl *tikvImplement) Get(req *endpoint.Request) (*httpc.Response, error) {
 	return impl.Client.
-		WithTimeout(defaultTimeout).
-		Get(req.Host, req.Port, buildRelativeURI(req.Path, req.Query))
+		WithTimeout(req.Timeout).
+		Get(req.Host, req.Port, buildRelativeURI(req.Path(), req.Query()))
+}
+
+func (impl *tikvImplement) CheckDest(host string, port int) (bool, error) {
+	info, _, err := topology.FetchStoreTopology(impl.PDClient)
+	if err != nil {
+		return false, err
+	}
+	return funk.Contains(info, func(item topology.StoreInfo) bool {
+		return item.IP == host && item.StatusPort == uint(port)
+	}), nil
 }
 
 type tiflashImplement struct {
 	fx.In
-	Client *tiflash.Client
+	Client   *tiflash.Client
+	PDClient *pd.Client
 }
 
 func (impl *tiflashImplement) Get(req *endpoint.Request) (*httpc.Response, error) {
 	return impl.Client.
-		WithTimeout(defaultTimeout).
-		Get(req.Host, req.Port, buildRelativeURI(req.Path, req.Query))
+		WithTimeout(req.Timeout).
+		Get(req.Host, req.Port, buildRelativeURI(req.Path(), req.Query()))
+}
+
+func (impl *tiflashImplement) CheckDest(host string, port int) (bool, error) {
+	_, info, err := topology.FetchStoreTopology(impl.PDClient)
```

Comment:
I believe this needs more careful treatment. For example, logically, the host restriction should be performed at the client side. In this way multiple services are benefited. However considering the scope I think you can separate them to multiple PRs.